### PR TITLE
[Source-Mssql] Assess and map all SQLServer supported datatypes to the correct JDBC and Airbyte types for Standard and CDC mode

### PR DESCRIPTION
**Pull request by nguyenaiden**
### Description
Addresses ticket #31296
Source-Mssql is currently still missing some SqlServer supported data types. This change revisited all of the [available types](https://learn.microsoft.com/en-us/sql/connect/jdbc/using-basic-data-types?view=sql-server-ver16) and map them to its corresponding JDBC types then to the appropriate Airbyte types.

*NOTE*: This is a breaking change as the following columns will now be discovered as a different type.
| Mssql type  | Current Airbyte Type | New Airbyte Type
|--|--|--|
| date  | string | date
| datetime  | string | timestamp
| smalldatetime  | string | timestamp 
| datetime2  | string | timestamp
| time  | string | time
| datetimeoffset  | string | timestamp with tz

### CODE REVIEW

**Comment by airbyte-oss-build-runner** [comment]
##  source-mssql-strict-encrypt test report (commit [`3af7b6b801`](https://github.com/airbytehq/airbyte/pull/31531/commits/3af7b6b8015e569913d77c6bbc037c8196a209ec)) - ❌

⏲️  Total pipeline duration: 302mn03s 

| Step                                                                        | Result   |
|:----------------------------------------------------------------------------|:---------|
| Build connector tar                                                         | ✅       |
| Java Connector Unit Tests                                                   | ❌       |
| Build source-mssql-strict-encrypt docker image for platform(s) linux/x86_64 | ✅       |
| Acceptance tests                                                            | ❌       |
| Java Connector Integration Tests                                            | ✅       |
| Validate metadata for source-mssql-strict-encrypt                           | ✅       |
| Connector version semver check                                              | ✅       |
| QA checks                                                                   | ❌       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/mssql-map-datatype/1697590830/3af7b6b8015e569913d77c6bbc037c8196a209ec/source-mssql-strict-encrypt/2.0.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:3af7b6b8015e569913d77c6bbc037c8196a209ec)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mssql-strict-encrypt test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-mssql test report (commit [`3af7b6b801`](https://github.com/airbytehq/airbyte/pull/31531/commits/3af7b6b8015e569913d77c6bbc037c8196a209ec)) - ❌

⏲️  Total pipeline duration: 13mn32s 

| Step                                                         | Result   |
|:-------------------------------------------------------------|:---------|
| Build connector tar                                          | ✅       |
| Java Connector Unit Tests                                    | ❌       |
| Build source-mssql docker image for platform(s) linux/x86_64 | ✅       |
| Acceptance tests                                             | ✅       |
| Java Connector Integration Tests                             | ✅       |
| Validate metadata for source-mssql                           | ✅       |
| Connector version semver check                               | ✅       |
| Connector version increment check                            | ✅       |
| QA checks                                                    | ❌       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/mssql-map-datatype/1697590830/3af7b6b8015e569913d77c6bbc037c8196a209ec/source-mssql/3.0.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:3af7b6b8015e569913d77c6bbc037c8196a209ec)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mssql test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-mssql-strict-encrypt test report (commit [`d14688c6d7`](https://github.com/airbytehq/airbyte/pull/31531/commits/d14688c6d71661ea44cbd5078630a28cb7461eae)) - ❌

⏲️  Total pipeline duration: 301mn31s 

| Step                                                                        | Result   |
|:----------------------------------------------------------------------------|:---------|
| Build connector tar                                                         | ✅       |
| Java Connector Unit Tests                                                   | ❌       |
| Build source-mssql-strict-encrypt docker image for platform(s) linux/x86_64 | ✅       |
| Acceptance tests                                                            | ❌       |
| Java Connector Integration Tests                                            | ✅       |
| Validate metadata for source-mssql-strict-encrypt                           | ✅       |
| Connector version semver check                                              | ✅       |
| QA checks                                                                   | ❌       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/mssql-map-datatype/1697662905/d14688c6d71661ea44cbd5078630a28cb7461eae/source-mssql-strict-encrypt/2.0.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:d14688c6d71661ea44cbd5078630a28cb7461eae)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mssql-strict-encrypt test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-mssql test report (commit [`d14688c6d7`](https://github.com/airbytehq/airbyte/pull/31531/commits/d14688c6d71661ea44cbd5078630a28cb7461eae)) - ❌

⏲️  Total pipeline duration: 13mn36s 

| Step                                                         | Result   |
|:-------------------------------------------------------------|:---------|
| Build connector tar                                          | ✅       |
| Java Connector Unit Tests                                    | ❌       |
| Build source-mssql docker image for platform(s) linux/x86_64 | ✅       |
| Acceptance tests                                             | ✅       |
| Java Connector Integration Tests                             | ✅       |
| Validate metadata for source-mssql                           | ✅       |
| Connector version semver check                               | ✅       |
| Connector version increment check                            | ✅       |
| QA checks                                                    | ❌       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/mssql-map-datatype/1697662905/d14688c6d71661ea44cbd5078630a28cb7461eae/source-mssql/3.0.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:d14688c6d71661ea44cbd5078630a28cb7461eae)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mssql test
```

**Review by prateekmukhedkar — COMMENTED** [review]
Can you check on the failing unit tests?

**Comment by airbyte-oss-build-runner** [comment]
##  source-mssql test report (commit [`47b9162d27`](https://github.com/airbytehq/airbyte/pull/31531/commits/47b9162d27fef5bfce9a98026c7097bc68818e50)) - ❌

⏲️  Total pipeline duration: 14mn04s 

| Step                                                         | Result   |
|:-------------------------------------------------------------|:---------|
| Build connector tar                                          | ✅       |
| Java Connector Unit Tests                                    | ✅       |
| Build source-mssql docker image for platform(s) linux/x86_64 | ✅       |
| Acceptance tests                                             | ✅       |
| Java Connector Integration Tests                             | ✅       |
| Validate metadata for source-mssql                           | ✅       |
| Connector version semver check                               | ✅       |
| Connector version increment check                            | ✅       |
| QA checks                                                    | ❌       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/mssql-map-datatype/1697748537/47b9162d27fef5bfce9a98026c7097bc68818e50/source-mssql/3.0.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:47b9162d27fef5bfce9a98026c7097bc68818e50)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mssql test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-mssql-strict-encrypt test report (commit [`2e6248ac2d`](https://github.com/airbytehq/airbyte/pull/31531/commits/2e6248ac2d950e52df6f85db973388794273aeb5)) - ❌

⏲️  Total pipeline duration: 302mn28s 

| Step                                                                        | Result   |
|:----------------------------------------------------------------------------|:---------|
| Build connector tar                                                         | ✅       |
| Java Connector Unit Tests                                                   | ❌       |
| Build source-mssql-strict-encrypt docker image for platform(s) linux/x86_64 | ✅       |
| Acceptance tests                                                            | ❌       |
| Java Connector Integration Tests                                            | ✅       |
| Validate metadata for source-mssql-strict-encrypt                           | ✅       |
| Connector version semver check                                              | ✅       |
| QA checks                                                                   | ❌       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/mssql-map-datatype/1697749508/2e6248ac2d950e52df6f85db973388794273aeb5/source-mssql-strict-encrypt/3.0.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:2e6248ac2d950e52df6f85db973388794273aeb5)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mssql-strict-encrypt test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-mssql test report (commit [`2e6248ac2d`](https://github.com/airbytehq/airbyte/pull/31531/commits/2e6248ac2d950e52df6f85db973388794273aeb5)) - ❌

⏲️  Total pipeline duration: 14mn02s 

| Step                                                         | Result   |
|:-------------------------------------------------------------|:---------|
| Build connector tar                                          | ✅       |
| Java Connector Unit Tests                                    | ✅       |
| Build source-mssql docker image for platform(s) linux/x86_64 | ✅       |
| Acceptance tests                                             | ✅       |
| Java Connector Integration Tests                             | ✅       |
| Validate metadata for source-mssql                           | ✅       |
| Connector version semver check                               | ✅       |
| Connector version increment check                            | ✅       |
| QA checks                                                    | ❌       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/mssql-map-datatype/1697749508/2e6248ac2d950e52df6f85db973388794273aeb5/source-mssql/3.0.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:2e6248ac2d950e52df6f85db973388794273aeb5)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mssql test
```

**Review by akashkulk — COMMENTED** [review]
Are these all the data type changes we foresee making? Asking only because if there are more, we should include all of them in a single breaking change

**Review by akashkulk — COMMENTED** [review]
For airbyte types can you specify timestamp w/or without timezone? 
And I think datetime sql servery type is also missing?

**Review by akashkulk — COMMENTED** [review]
I think datetimeoffset should be timestamp w/ tz and datetime, datetime2 and smalldatetime should be timezone w/out tz. WDYT?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java [hunk 1]
```diff
@@ -88,6 +89,28 @@ public JDBCType getDatabaseFieldType(final JsonNode field) {
...
       }
+
+      /**
+       * Per Microsoft SQL Server docs java.sql.Timestamp values can no longer be used to compare values
+       * from a datetime column starting from SQL Server 2016. This limitation is due to a server-side
+       * change that converts datetime to datetime2 differently, resulting in non-equal values. The
+       * workaround to this issue is to either change datetime columns to datetime2(3), use String instead
+       * of java.sql.Timestamp, or change database compatibility level to 120 or lower.
+       *
+       * So, we will treat datetime columns as VARCHAR
```
Comment (nguyenaiden) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java:100]: @akashkulk this is where I'm handling `datetime`.

**Comment by nguyenaiden** [comment]
> I think datetimeoffset should be timestamp w/ tz and datetime, datetime2 and smalldatetime should be timezone w/out tz. WDYT?

@akashkulk I believe I'm already doing that no? Aside from `datetime` which we're handling as string - see comment as of why I'm handling it this way, the types that you mentioned above is being asserted in the tests [AbstractMssqlSourceDatatypeTest.java](https://github.com/airbytehq/airbyte/pull/31531/files#diff-aede9bf8b44df13722a78f96f982e3bb549930c78d2b06d74fafd2f713a7f403)

**Review by akashkulk — APPROVED** [review]

File: airbyte-integrations/connectors/source-mssql/metadata.yaml [hunk 2]
```diff
@@ -29,6 +29,9 @@ data:
     - language:python
   releases:
     breakingChanges:
+      3.0.0:
+        message: "Remapped columns of types: date, datetime, datetime2, datetimeoffset, smalldatetime, and time from `String` to their appropriate Airbyte types. Customers whose streams have columns with the affected datatype must refresh their schema and reset their data."
+        upgradeDeadline: "2023-10-31"
```
Comment (katmarkham) [inline airbyte-integrations/connectors/source-mssql/metadata.yaml:34]: the upgrade deadline needs to be at least 7 days out.  This date is in the past

**Review by katmarkham — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-mssql/metadata.yaml (hunk 2 above)
Comment (nguyenaiden) [reply airbyte-integrations/connectors/source-mssql/metadata.yaml:34]: This is just a placeholder. I'm going to update it to 1 week from the day we send out comms.

**Review by katmarkham — APPROVED** [review]

**Pull request merged**